File: implementations/python/src/foilbench_py/benchmark/runner.py
"""Matched-physical-run benchmark harness."""

import csv
import json
import os
import platform
import statistics
import subprocess
import time
from dataclasses import dataclass, replace
from pathlib import Path
from typing import cast

import numpy as np
import psutil

from foilbench_py.benchmark.artifact import solver_configuration, validate_result_semantics
from foilbench_py.core._schema_adapter import validate_json
from foilbench_py.core.geometry import NacaFoil
from foilbench_py.core.metrics import analyze_wake_probe
from foilbench_py.core.models import NumericalFailure, Scenario, StepReport
from foilbench_py.core.protocol import FlowSolver
from foilbench_py.core.scenario import find_repo_root, load_scenario
from foilbench_py.core.state_io import save_canonical_state
from foilbench_py.solvers.factory import create_solver


@dataclass(frozen=True, slots=True)
class BenchmarkMatrix:
    id: str
    scenario_path: Path
    solvers: tuple[str, ...]
    resolutions: tuple[tuple[int, int], ...]
    duration: float
    repetitions: int
    save_snapshots: bool


def _json_object(path: Path) -> dict[str, object]:
    value = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(value, dict):
        raise TypeError(f"{path} must contain a JSON object")
    return cast(dict[str, object], value)


def load_matrix(path: str | Path) -> BenchmarkMatrix:
    matrix_path = Path(path).resolve()
    raw = _json_object(matrix_path)
    root = find_repo_root(matrix_path)
    validate_json(raw, _json_object(root / "spec" / "benchmark-matrix.schema.json"))
    scenario_path = root / str(raw["scenario"])
    raw_resolutions = cast(list[list[int]], raw["resolutions"])
    return BenchmarkMatrix(
        id=str(raw["id"]),
        scenario_path=scenario_path,
        solvers=tuple(str(item) for item in cast(list[str], raw["solvers"])),
        resolutions=tuple((int(item[0]), int(item[1])) for item in raw_resolutions),
        duration=float(cast(float, raw["duration"])),
        repetitions=int(cast(int, raw["repetitions"])),
        save_snapshots=cast(bool, raw["save_snapshots"]),
    )


def _git_commit(root: Path) -> str:
    try:
        completed = subprocess.run(
            ["git", "rev-parse", "HEAD"],
            cwd=root,
            check=True,
            capture_output=True,
            text=True,
        )
        return completed.stdout.strip()
    except (OSError, subprocess.CalledProcessError):
        return "unknown"


def _machine() -> dict[str, object]:
    return {
        "platform": platform.platform(),
        "python": platform.python_version(),
        "processor": platform.processor(),
        "logical_cpus": os.cpu_count(),
        "numpy": np.__version__,
    }


def _percentile_95(values: list[float]) -> float:
    if len(values) == 1:
        return values[0]
    return float(np.percentile(np.asarray(values), 95.0))


type JsonScalar = float | int | bool | str | None


def _json_scalar(value: object) -> JsonScalar:
    if isinstance(value, np.generic):
        return _json_scalar(cast(object, value.item()))
    if isinstance(value, float) and not np.isfinite(value):
        raise ValueError("benchmark evidence values must be finite")
    if value is None or isinstance(value, (bool, int, float, str)):
        return value
    raise TypeError(f"benchmark evidence value {value!r} is not a JSON scalar")


def _step_artifact(report: StepReport | None) -> dict[str, object] | None:
    if report is None:
        return None
    return {
        "requested_dt": report.requested_dt,
        "advanced_dt": report.advanced_dt,
        "substeps": report.substeps,
        "max_speed": report.max_speed,
        "state_revision": report.state_revision,
        "evidence": {key: _json_scalar(value) for key, value in report.evidence.items()},
        "warnings": list(report.warnings),
    }


def _failure_artifact(error: Exception) -> dict[str, object]:
    if isinstance(error, NumericalFailure):
        return {
            "kind": "numerical",
            "reason": error.reason,
            "stage": error.stage,
            "message": str(error),
            "evidence": {
                key: _json_scalar(value) for key, value in error.evidence.items()
            },
        }
    return {
        "kind": "unexpected",
        "reason": None,
        "stage": None,
        "message": f"{type(error).__name__}: {error}",
        "evidence": {},
    }


def recovery_window(scenario: Scenario) -> tuple[float, float] | None:
    """Return the baseline end and completed-return time for an angle excursion."""
    initial_angle = scenario.controls[0].angle_degrees
    final_angle = scenario.controls[-1].angle_degrees
    if not np.isclose(initial_angle, final_angle):
        return None
    changed = [
        index
        for index, keyframe in enumerate(scenario.controls)
        if not np.isclose(keyframe.angle_degrees, initial_angle)
    ]
    if not changed:
        return None
    first_changed = changed[0]
    last_changed = changed[-1]
    if first_changed == 0 or last_changed + 1 >= len(scenario.controls):
        return None
    baseline_end = scenario.controls[first_changed - 1].time
    recovery_start = scenario.controls[last_changed + 1].time
    if baseline_end >= recovery_start or recovery_start >= scenario.duration:
        return None
    return baseline_end, recovery_start


def run_matrix(
    matrix_path: str | Path,
    output_root: str | Path | None = None,
) -> Path:
    matrix = load_matrix(matrix_path)
    root = find_repo_root(Path(matrix_path))
    destination = (
        Path(output_root)
        if output_root is not None
        else root / "results" / matrix.id / time.strftime("%Y%m%d-%H%M%S")
    )
    destination.mkdir(parents=True, exist_ok=True)
    scenario_base = load_scenario(matrix.scenario_path)
    result_schema = _json_object(root / "spec" / "result.schema.json")
    rows: list[dict[str, object]] = []
    process = psutil.Process()

    for resolution in matrix.resolutions:
        scenario = replace(
            scenario_base,
            domain=replace(scenario_base.domain, resolution=resolution),
            duration=matrix.duration,
        )
        geometry = NacaFoil(scenario.foil)
        for solver_id in matrix.solvers:
            for repetition in range(matrix.repetitions):
                initialization_seconds = 0.0
                cold_step_seconds = 0.0
                solver: FlowSolver | None = None
                elapsed_simulated = 0.0
                step_seconds: list[float] = []
                total_substeps = 0
                peak_rss = process.memory_info().rss
                warnings: list[str] = []
                wake_probe: list[float] = []
                recovery_times = recovery_window(scenario)
                recovery_baseline: tuple[float, float] | None = None
                recovery_elapsed: float | None = None
                last_report: StepReport | None = None
                diagnostic_revision: int | None = None
                failure: dict[str, object] | None = None
                diagnostic_values: dict[str, float] = {}
                probe_point = np.asarray(
                    [
                        [
                            min(
                                scenario.foil.pivot[0] + 1.5 * scenario.foil.chord,
                                scenario.domain.bounds[0][1] - 0.5 * scenario.domain.dx,
                            ),
                            scenario.foil.pivot[1],
                        ]
                    ],
                    dtype=scenario.dtype,
                )
                success = True
                try:
                    cold_solver = create_solver(solver_id)
                    init_start = time.perf_counter()
                    cold_solver.initialize(scenario, geometry, scenario.seed)
                    initialization_seconds = time.perf_counter() - init_start

                    warm_dt = min(scenario.output_dt, scenario.duration)
                    warm_control = scenario.control_at(warm_dt)
                    cold_step_start = time.perf_counter()
                    cold_solver.advance(warm_control, warm_dt)
                    cold_step_seconds = time.perf_counter() - cold_step_start

                    # Reinitialize after warming process-global compiled kernels so the
                    # measured physical run starts from the declared scenario state.
                    solver = create_solver(solver_id)
                    solver.initialize(scenario, geometry, scenario.seed)
                    while elapsed_simulated < scenario.duration - 1.0e-12:
                        dt = min(scenario.output_dt, scenario.duration - elapsed_simulated)
                        control = scenario.control_at(elapsed_simulated + dt)
                        started = time.perf_counter()
                        report = solver.advance(control, dt)
                        last_report = report
                        step_seconds.append(time.perf_counter() - started)
                        elapsed_simulated += report.advanced_dt
                        total_substeps += report.substeps
                        warnings.extend(report.warnings)
                        peak_rss = max(peak_rss, process.memory_info().rss)
                        if elapsed_simulated >= 0.5 * scenario.duration:
                            wake_probe.append(float(solver.sample_velocity(probe_point)[0, 1]))
                        if recovery_times is not None:
                            baseline_end, recovery_start = recovery_times
                            crossed_baseline = (
                                recovery_baseline is None
                                and elapsed_simulated >= baseline_end
                            )
                            observing_recovery = (
                                recovery_baseline is not None
                                and recovery_elapsed is None
                                and elapsed_simulated >= recovery_start
                            )
                            if crossed_baseline or observing_recovery:
                                transient = solver.diagnostics().values
                                wake = transient["wake_width"]
                                recirculation = transient["recirculation_area"]
                                if crossed_baseline:
                                    recovery_baseline = (wake, recirculation)
                                elif recovery_baseline is not None:
                                    wake_tolerance = max(
                                        1.25 * recovery_baseline[0],
                                        2.0 * scenario.domain.dy,
                                    )
                                    recirculation_tolerance = max(
                                        1.25 * recovery_baseline[1],
                                        2.0 * scenario.domain.dx * scenario.domain.dy,
                                    )
                                    if (
                                        wake <= wake_tolerance
                                        and recirculation <= recirculation_tolerance
                                    ):
                                        recovery_elapsed = elapsed_simulated - recovery_start
                    diagnostics = solver.diagnostics()
                    if diagnostics.state_revision != solver.state_revision:
                        raise RuntimeError("benchmark diagnostics describe a stale state revision")
                    diagnostic_revision = diagnostics.state_revision
                    warnings.extend(diagnostics.warnings)
                    diagnostic_values = dict(diagnostics.values)
                    if len(wake_probe) >= 8:
                        spectrum = analyze_wake_probe(
                            np.asarray(wake_probe, dtype=np.float64),
                            scenario.output_dt,
                            scenario.foil.chord,
                            max(float(np.linalg.norm(scenario.freestream)), 1.0e-12),
                        )
                        diagnostic_values.update(
                            {
                                "wake_probe_samples": float(spectrum.sample_count),
                                "wake_frequency_resolution": spectrum.frequency_resolution,
                                "wake_transverse_rms": spectrum.transverse_rms,
                                "wake_mixing_index": (
                                    spectrum.transverse_rms / scenario.reference_speed
                                ),
                                "wake_dominant_frequency": spectrum.dominant_frequency,
                                "wake_strouhal_number": spectrum.strouhal_number,
                                "wake_dominant_power_fraction": (
                                    spectrum.dominant_power_fraction
                                ),
                            }
                        )
                    if recovery_times is not None and recovery_baseline is not None:
                        baseline_end, recovery_start = recovery_times
                        observed = recovery_elapsed is not None
                        diagnostic_values.update(
                            {
                                "recovery_baseline_time": baseline_end,
                                "recovery_start_time": recovery_start,
                                "recovery_observed": float(observed),
                                "recovery_elapsed": (
                                    recovery_elapsed
                                    if recovery_elapsed is not None
                                    else scenario.duration - recovery_start
                                ),
                            }
                        )
                        if not observed:
                            warnings.append(
                                "wake recovery was not observed; recovery_elapsed is right-censored"
                            )
                except Exception as error:
                    success = False
                    warnings.append(f"{type(error).__name__}: {error}")
                    diagnostic_values = {}
                    diagnostic_revision = None
                    failure = _failure_artifact(error)

                total_wall = sum(step_seconds)
                median = statistics.median(step_seconds) if step_seconds else 0.0
                p95 = _percentile_95(step_seconds) if step_seconds else 0.0
                cell_updates_per_second = (
                    scenario.domain.nx * scenario.domain.ny * total_substeps / total_wall
                    if total_wall > 0.0
                    else 0.0
                )
                particle_count = diagnostic_values.get("particle_count", 0.0)
                particle_updates_per_second = (
                    particle_count * total_substeps / total_wall if total_wall > 0.0 else 0.0
                )
                result: dict[str, object] = {
                    "schema_version": 1,
                    "contract_id": "foilbench-phase2-v1",
                    "contract_revision": 4,
                    "benchmark_matrix_id": matrix.id,
                    "scenario_id": scenario.id,
                    "repetition": repetition + 1,
                    "language": "python",
                    "solver": solver_id,
                    "git_commit": _git_commit(root),
                    "machine": _machine(),
                    "precision": scenario.precision,
                    "resolution": list(scenario.domain.resolution),
                    "bounds": [list(pair) for pair in scenario.domain.bounds],
                    "periodic_axes": list(scenario.domain.periodic_axes),
                    "reynolds": scenario.reynolds,
                    "effective_reynolds": diagnostic_values.get(
                        "effective_reynolds", scenario.reynolds
                    ),
                    "solver_configuration": solver_configuration(scenario),
                    "freestream": list(scenario.freestream),
                    "foil": {
                        "naca": scenario.foil.naca,
                        "chord": scenario.foil.chord,
                        "pivot": list(scenario.foil.pivot),
                    },
                    "control_history": [
                        {"time": keyframe.time, "angle_degrees": keyframe.angle_degrees}
                        for keyframe in scenario.controls
                    ],
                    "requested_duration": matrix.duration,
                    "simulated_duration": elapsed_simulated,
                    "output_dt": scenario.output_dt,
                    "seed": scenario.seed,
                    "initialization_seconds": initialization_seconds,
                    "cold_step_seconds": cold_step_seconds,
                    "step_seconds": step_seconds,
                    "median_step_seconds": median,
                    "p95_step_seconds": p95,
                    "simulated_seconds_per_wall_second": (
                        elapsed_simulated / total_wall if total_wall > 0.0 else 0.0
                    ),
                    "cell_updates_per_second": cell_updates_per_second,
                    "particle_updates_per_second": particle_updates_per_second,
                    "peak_rss_bytes": peak_rss,
                    "memory_measurement": "rss",
                    "runtime_startup_seconds": None,
                    "worker_startup_seconds": None,
                    "substeps": total_substeps,
                    "final_state_revision": 0 if solver is None else solver.state_revision,
                    "diagnostic_state_revision": diagnostic_revision,
                    "last_step": _step_artifact(last_report),
                    "diagnostics": diagnostic_values,
                    "success": success,
                    "failure": failure,
                    "warnings": sorted(set(warnings)),
                }
                validate_json(result, result_schema)
                validate_result_semantics(result)
                stem = f"{solver_id}-{resolution[0]}x{resolution[1]}-r{repetition + 1}"
                result_path = destination / f"{stem}.json"
                result_path.write_text(
                    json.dumps(result, indent=2, allow_nan=False), encoding="utf-8"
                )
                if matrix.save_snapshots and success and solver is not None:
                    save_canonical_state(solver.export_state(), destination / f"{stem}-state")
                rows.append(
                    {
                        "solver": solver_id,
                        "resolution": f"{resolution[0]}x{resolution[1]}",
                        "repetition": repetition + 1,
                        "initialization_seconds": initialization_seconds,
                        "cold_step_seconds": cold_step_seconds,
                        "median_step_seconds": median,
                        "p95_step_seconds": p95,
                        "simulated_seconds_per_wall_second": result[
                            "simulated_seconds_per_wall_second"
                        ],
                        "cell_updates_per_second": cell_updates_per_second,
                        "particle_updates_per_second": particle_updates_per_second,
                        "peak_rss_bytes": peak_rss,
                        "success": success,
                    }
                )

    if rows:
        with (destination / "summary.csv").open("w", newline="", encoding="utf-8") as handle:
            writer = csv.DictWriter(handle, fieldnames=list(rows[0]))
            writer.writeheader()
            writer.writerows(rows)
    return destination

File: implementations/typescript/src/benchmark/runner.ts
import Ajv2020 from "ajv/dist/2020.js";
import {chromium} from "playwright";
import {execFileSync} from "node:child_process";
import {mkdir, readFile, writeFile} from "node:fs/promises";
import {cpus, platform, release} from "node:os";
import {dirname, isAbsolute, join, resolve} from "node:path";
import {fileURLToPath} from "node:url";
import {createServer} from "vite";
import type {Scenario, SolverId, StepReport} from "../core/contracts.js";
import {parseScenario, validateDocument} from "../core/scenario.js";
import type {BenchmarkMatrix, BrowserRunRequest, BrowserRunResult} from "./types.js";
import type {BrowserCanonicalSnapshot} from "./types.js";

interface RawMatrix {schema_version: 1; id: string; scenario: string; solvers: SolverId[]; resolutions: [number, number][]; duration: number; repetitions: number; save_snapshots: boolean}
function repositoryRoot(): string { return resolve(dirname(fileURLToPath(import.meta.url)), "../../../.."); }
async function json(path: string): Promise<unknown> { return JSON.parse(await readFile(path, "utf8")) as unknown; }
function percentile(values: readonly number[], fraction: number): number { if (values.length === 0) return 0; const sorted = [...values].sort((a, b) => a - b); const position = fraction * (sorted.length - 1); const lower = Math.floor(position); const upper = Math.ceil(position); const a = sorted[lower] ?? 0; const b = sorted[upper] ?? a; return a + (b - a) * (position - lower); }
function stepArtifact(report: StepReport | null): Record<string, unknown> | null { return report === null ? null : {requested_dt: report.requestedDt, advanced_dt: report.advancedDt, substeps: report.substeps, max_speed: report.maxSpeed, state_revision: report.stateRevision, evidence: report.evidence, warnings: report.warnings}; }
const PHYSICAL_IDENTITY_FIELDS = ["bounds", "periodic_axes", "reynolds", "effective_reynolds", "solver_configuration", "freestream", "foil", "control_history", "requested_duration", "output_dt", "seed"] as const;

function canonicalize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonicalize);
  if (value !== null && typeof value === "object") {
    return Object.fromEntries(Object.entries(value as Record<string, unknown>).sort(([left], [right]) => left.localeCompare(right)).map(([key, child]) => [key, canonicalize(child)]));
  }
  return value;
}

function physicalIdentity(value: Readonly<Record<string, unknown>>): string {
  return JSON.stringify(canonicalize(Object.fromEntries(PHYSICAL_IDENTITY_FIELDS.map((field) => [field, value[field]]))));
}

function solverConfiguration(scenario: Scenario): Record<string, unknown> {
  return {
    initial_condition: scenario.solverOptions.initialCondition ?? "freestream",
    stable_advection: scenario.solverOptions.stableAdvection ?? "maccormack",
    stable_face_advection: scenario.solverOptions.stableFaceAdvection ?? false,
    stable_cfl: scenario.solverOptions.stableCfl ?? 0.7,
    pressure_tolerance: scenario.solverOptions.pressureTolerance ?? 1e-5,
    pressure_max_iterations: scenario.solverOptions.pressureMaxIterations ?? 640,
    pic_flip_blend: scenario.solverOptions.picFlipBlend ?? 0.95,
    pic_population_interval: scenario.solverOptions.picPopulationInterval ?? 8,
    pic_cfl: scenario.solverOptions.picCfl ?? 0.75,
  };
}

function requireFinite(value: unknown, path = "result"): void {
  if (typeof value === "number" && !Number.isFinite(value)) throw new TypeError(`${path} contains a non-finite number`);
  if (Array.isArray(value)) value.forEach((child, index) => requireFinite(child, `${path}[${String(index)}]`));
  else if (value !== null && typeof value === "object") for (const [key, child] of Object.entries(value)) requireFinite(child, `${path}.${key}`);
}

export function validateResultSemantics(value: Readonly<Record<string, unknown>>): void {
  requireFinite(value);
  const success = value["success"] === true;
  const lastStep = value["last_step"];
  const steps = value["step_seconds"];
  const finalRevision = value["final_state_revision"];
  const precision = value["precision"];
  const requested = Number(value["requested_duration"]);
  const simulated = Number(value["simulated_duration"]);
  const tolerance = (precision === "float32" ? 1e-6 : 1e-12) * Math.max(1, Math.abs(requested));
  if (success) {
    if (value["failure"] !== null || lastStep === null || typeof lastStep !== "object" || !Array.isArray(steps) || steps.length === 0) throw new TypeError("successful benchmark result lacks completed-step semantics");
    const stepRevision = (lastStep as Record<string, unknown>)["state_revision"];
    if (value["diagnostic_state_revision"] !== finalRevision || stepRevision !== finalRevision) throw new TypeError("successful benchmark result contains stale revision evidence");
    if (Math.abs(simulated - requested) > tolerance) throw new TypeError("successful benchmark result did not complete requested duration");
  } else if (value["failure"] === null || typeof value["failure"] !== "object") throw new TypeError("failed benchmark result lacks structured failure evidence");
  if ((value["memory_measurement"] === "unavailable") !== (value["peak_rss_bytes"] === null)) throw new TypeError("memory measurement kind and RSS value disagree");
}

export async function loadMatrix(path: string): Promise<BenchmarkMatrix> { const root = repositoryRoot(); const document = await json(isAbsolute(path) ? path : join(root, path)); validateDocument(document, await json(join(root, "spec/benchmark-matrix.schema.json")) as object); const raw = document as RawMatrix; return {id: raw.id, scenario: raw.scenario, solvers: raw.solvers, resolutions: raw.resolutions, duration: raw.duration, repetitions: raw.repetitions, saveSnapshots: raw.save_snapshots}; }

export async function runBrowserMatrix(matrixPath: string, outputRoot?: string): Promise<string> { const root = repositoryRoot(); const matrix = await loadMatrix(matrixPath); const scenario = parseScenario(await json(join(root, matrix.scenario)), await json(join(root, "spec/scenario.schema.json")) as object); const resultSchema = await json(join(root, "spec/result.schema.json")) as object; const validator = new Ajv2020({strict: true, allErrors: true}).compile(resultSchema); const destination = outputRoot === undefined ? join(root, "results", matrix.id, new Date().toISOString().replaceAll(/[:.]/g, "-")) : isAbsolute(outputRoot) ? outputRoot : join(root, outputRoot); await mkdir(destination, {recursive: true}); const server = await createServer({root: resolve(root, "implementations/typescript"), server: {host: "127.0.0.1", port: 0}}); await server.listen(); const address = server.httpServer?.address(); if (address === null || typeof address === "string" || address === undefined) throw new Error("Vite benchmark server did not bind"); const runtimeStarted = performance.now(); const browser = await chromium.launch({headless: true}); const runtimeStartupSeconds = (performance.now() - runtimeStarted) / 1000; const page = await browser.newPage(); await page.goto(`http://127.0.0.1:${String(address.port)}/benchmark.html`); const rows: string[][] = [];
  try { for (const resolution of matrix.resolutions) for (const solverId of matrix.solvers) for (let repetition = 1; repetition <= matrix.repetitions; repetition += 1) { const selected: Scenario = {...scenario, domain: {...scenario.domain, resolution}}; const request: BrowserRunRequest = {scenario: selected, solverId, duration: matrix.duration}; const response = await page.evaluate(async (value) => { const browser = globalThis as unknown as {runFoilBench(request: BrowserRunRequest): Promise<{workerStartupSeconds: number; result: BrowserRunResult}>}; return browser.runFoilBench(value); }, request); const run: BrowserRunResult = response.result; const totalWall = run.stepSeconds.reduce((left, right) => left + right, 0); const median = percentile(run.stepSeconds, 0.5); const p95 = percentile(run.stepSeconds, 0.95); const cells = resolution[0] * resolution[1]; const particleCount = run.diagnostics["particle_count"] ?? 0; const effectiveReynolds = Number(run.diagnostics["effective_reynolds"] ?? run.lastStep?.evidence["effective_reynolds"] ?? selected.reynolds); const result = {schema_version: 1, contract_id: "foilbench-phase2-v1", contract_revision: 4, benchmark_matrix_id: matrix.id, scenario_id: selected.id, repetition, language: "typescript", solver: solverId, git_commit: gitCommit(root), machine: {platform: `${platform()} ${release()}`, node: process.version, browser: "chromium", logical_cpus: cpus().length}, precision: selected.precision, resolution: [...resolution], bounds: selected.domain.bounds, periodic_axes: selected.domain.periodicAxes, reynolds: selected.reynolds, effective_reynolds: effectiveReynolds, solver_configuration: solverConfiguration(selected), freestream: selected.freestream, foil: selected.foil, control_history: selected.controls.map((control) => ({time: control.time, angle_degrees: control.angleDegrees})), requested_duration: matrix.duration, simulated_duration: run.simulatedSeconds, output_dt: selected.outputDt, seed: selected.seed, initialization_seconds: run.initializationSeconds, cold_step_seconds: run.coldStepSeconds, step_seconds: run.stepSeconds, median_step_seconds: median, p95_step_seconds: p95, simulated_seconds_per_wall_second: totalWall > 0 ? run.simulatedSeconds / totalWall : 0, cell_updates_per_second: totalWall > 0 ? cells * run.substeps / totalWall : 0, particle_updates_per_second: totalWall > 0 ? particleCount * run.substeps / totalWall : 0, peak_rss_bytes: null, memory_measurement: "unavailable", runtime_startup_seconds: runtimeStartupSeconds, worker_startup_seconds: response.workerStartupSeconds, substeps: run.substeps, final_state_revision: run.finalStateRevision, diagnostic_state_revision: run.diagnosticStateRevision, last_step: stepArtifact(run.lastStep), diagnostics: run.diagnostics, success: run.success, failure: run.failure, warnings: run.warnings}; if (!validator(result)) throw new TypeError(new Ajv2020().errorsText(validator.errors)); validateResultSemantics(result); const stem = `${solverId}-${String(resolution[0])}x${String(resolution[1])}-r${String(repetition)}`; await writeFile(join(destination, `${stem}.json`), JSON.stringify(result, null, 2), "utf8"); if (matrix.saveSnapshots && run.success && run.snapshot !== null) await saveCanonicalSnapshot(join(destination, `${stem}-state`), run.snapshot, solverId, await json(join(root, "spec/canonical-manifest.schema.json")) as object); rows.push([solverId, `${String(resolution[0])}x${String(resolution[1])}`, String(repetition), String(median), String(p95), String(result.simulated_seconds_per_wall_second), String(result.success)]); } }
  finally { await browser.close(); await server.close(); }
  await writeFile(join(destination, "summary.csv"), [["solver", "resolution", "repetition", "median_step_seconds", "p95_step_seconds", "simulated_seconds_per_wall_second", "success"], ...rows].map((row) => row.join(",")).join("\n") + "\n", "utf8"); return destination; }
function gitCommit(root: string): string { try { return execFileSync("git", ["rev-parse", "HEAD"], {cwd: root, encoding: "utf8"}).trim(); } catch { return "unknown"; } }

function encodeNpy(values: readonly number[], precision: "float32" | "float64", shape: readonly number[]): Buffer { const bytes = precision === "float32" ? 4 : 8; const descriptor = precision === "float32" ? "<f4" : "<f8"; const shapeText = shape.length === 1 ? `${String(shape[0])},` : shape.map(String).join(", "); const prefixLength = 10; let header = `{'descr': '${descriptor}', 'fortran_order': False, 'shape': (${shapeText}), }`; const padding = (16 - ((prefixLength + header.length + 1) % 16)) % 16; header += " ".repeat(padding) + "\n"; const output = Buffer.alloc(prefixLength + header.length + bytes * values.length); output[0] = 0x93; output.write("NUMPY", 1, "ascii"); output[6] = 1; output[7] = 0; output.writeUInt16LE(header.length, 8); output.write(header, prefixLength, "latin1"); const view = new DataView(output.buffer, output.byteOffset + prefixLength + header.length, bytes * values.length); for (let index = 0; index < values.length; index += 1) { const value = values[index] ?? 0; if (bytes === 4) view.setFloat32(index * bytes, value, true); else view.setFloat64(index * bytes, value, true); } return output; }

async function saveCanonicalSnapshot(directory: string, snapshot: BrowserCanonicalSnapshot, solverId: SolverId, schema: object): Promise<void> { await mkdir(directory, {recursive: true}); const [nx, ny] = snapshot.resolution; if (nx === undefined || ny === undefined) throw new RangeError("2D snapshot resolution missing"); const velocityMetadata = {file: "velocity.npy", axes: ["z", "y", "x", "component"], order: "C"}; const densityMetadata = snapshot.density === null ? null : {file: "density.npy", axes: ["z", "y", "x"], order: "C"}; const manifest = {schema_version: 1, dimension: 2, bounds: snapshot.bounds, resolution: snapshot.resolution, periodic_axes: snapshot.periodicAxes, time: snapshot.time, precision: snapshot.precision, angle_degrees: snapshot.angleDegrees, angular_velocity_degrees: snapshot.angularVelocityDegrees, source_language: "typescript", source_solver: solverId, velocity: velocityMetadata, density: densityMetadata}; validateDocument(manifest, schema); await writeFile(join(directory, "velocity.npy"), encodeNpy(snapshot.velocity, snapshot.precision, [1, ny, nx, 2])); if (snapshot.density !== null) await writeFile(join(directory, "density.npy"), encodeNpy(snapshot.density, snapshot.precision, [1, ny, nx])); await writeFile(join(directory, "manifest.json"), JSON.stringify(manifest, null, 2), "utf8"); }

export async function compareResults(directory: string): Promise<string> {
  const root = repositoryRoot();
  const selected = isAbsolute(directory) ? directory : join(root, directory);
  const files = await import("node:fs/promises").then(({readdir}) => readdir(selected, {recursive: true}));
  const schema = await json(join(root, "spec/result.schema.json")) as object;
  const validator = new Ajv2020({strict: true, allErrors: true}).compile(schema);
  const signatures = new Map<string, string>();
  const lines = ["language    solver              median ms      p95 ms    sim/wall success"];
  for (const file of files) {
    if (!file.endsWith(".json")) continue;
    const value = await json(join(selected, file)) as Record<string, unknown>;
    const solver = value["solver"];
    if (typeof solver !== "string") continue;
    if (!validator(value)) throw new TypeError(new Ajv2020().errorsText(validator.errors));
    validateResultSemantics(value);
    const key = JSON.stringify([value["benchmark_matrix_id"], value["scenario_id"], value["precision"], value["resolution"], value["solver"]]);
    const signature = physicalIdentity(value);
    const previous = signatures.get(key);
    if (previous !== undefined && previous !== signature) throw new Error("benchmark artifacts reuse a matrix/scenario/resolution identity with different physical inputs");
    signatures.set(key, signature);
    const language = typeof value["language"] === "string" ? value["language"] : "unknown";
    const median = (Number(value["median_step_seconds"]) * 1000).toFixed(3);
    const p95 = (Number(value["p95_step_seconds"]) * 1000).toFixed(3);
    const throughput = Number(value["simulated_seconds_per_wall_second"]).toFixed(3);
    lines.push(`${language.padEnd(12)}${solver.padEnd(20)}${median.padStart(10)}${p95.padStart(12)}${throughput.padStart(12)} ${String(value["success"])}`);
  }
  return lines.join("\n");
}
